### ISSUE TITLE
Synthetic lifecycle observable changes when slotted into native lifecycle

### ISSUE THREAD

**Issue by nolanlawson** [salesforce/lwc#4249]
_**TL;DR**: No matter what, there will be some observable difference if a synthetic-lifecycle component is slotted into a native-lifecycle component._

As demonstrated by #4248, we have some issues where we have incorrect behavior with a synthetic lifecycle component is slotted into a native lifecycle component. In particular, its `disconnectedCallback` may be invoked without a corresponding `connectedCallback` (for a true case of disconnect+reconnect).

We can fix these bugs, but I think there will still be cases where we can't perfectly emulate the old pure-synthetic-lifecycle behavior for synthetic-lifecycle components. For example:

```html
<!-- slottable.html -->
<template lwc:if={foo}>
  <slot></slot>
</template>
```

```js
// slottable.js
renderedCallback() {
  this.foo = true
}
```

```html
<!-- container -->
<x-slottable> <!-- native lifecycle -->
  <x-slottee></x-slottee> <!-- synthetic lifecycle -->
</x-slottable>
```

In this case, a synthetic-lifecycle component is being slotted into a native-lifecycle component. However, the outer component is changing its state within its `renderedCallback`, which affects how the inner component is mounted/unmounted. (I used `renderedCallback`, but I could have also used `disconnectedCallback`/`connectedCallback`.)

The point is that the outer component has changed its timing due to native lifecycle. Necessarily, the inner component will be unmounted/mounted based on the _native_ timing, not the synthetic timing it expects. So no matter what, there will be some observable difference if a synthetic-lifecycle component is slotted into a native-lifecycle component.
